File: problem-14/c/main.c
#include <stdio.h>
#include <stdint.h>

int collatz(uint64_t n) {
	// Base Case
	if (n == 1) {
		return 1;
	}

	// Next State
	uint64_t next = 0;
	if (n % 2 == 0) {
		next = n / 2;
	}
	else {
		next = (3 * n) + 1;
	}

	return collatz(next) + 1;
}

uint64_t solution(uint64_t limit) {
	// Prepare
	uint64_t idx = 0;
	int max = 0;

	// Search
	for (uint64_t i = 1; i < limit; i++) {
		int count = collatz(i);
		if (count > max) {
			max = count;
			idx = i;
		}
	}

	return idx;
}

int main() {
	printf("Solution #14: %ld\n", solution(1000000));
}

File: problem-14/rust/src/main.rs
fn main() {
    println!("Solution #14: {}", solution(1_000_000));
}

fn solution(n: u64) -> usize {
    return (0..n)
        .map(|x| Collatz::new(x).count())
        .enumerate()
        .max_by_key(|x| x.1)
        .unwrap()
        .0
}

struct Collatz {
    n: u64,
}

impl Collatz {
    fn new(n: u64) -> Collatz {
        return Collatz{
            n: n,
        }
    }
}

impl Iterator for Collatz {
    type Item = u64;

    fn next(&mut self) -> Option<u64> {
        // Default Case
        if self.n == 0 {
            return None
        }

        // Output Value
        let output = Some(self.n);

        // Update State
        self.n = match self.n {
            1 => 0,
            _ => match self.n % 2 {
                0 => self.n / 2,
                1 => (3 * self.n) + 1,
                _ => 1,
            },
        };
        
        return output
    }
}
